aten/src/ATen/core/custom_class.cpp

```diff
@@ -1,27 +1,49 @@
+#include <torch/custom_class.h>
+
 #include <ATen/core/jit_type.h>
-#include <torch/csrc/jit/api/custom_class.h>
 
 #include <atomic>
+#include <unordered_map>
 
 namespace torch {
 namespace jit {
 
-namespace {
+std::unordered_map<std::string, at::ClassTypePtr>& customClasses() {
+  static std::unordered_map<std::string, at::ClassTypePtr> customClasses;
+  return customClasses;
+}
+
+void registerCustomClass(at::ClassTypePtr class_type) {
+  TORCH_INTERNAL_ASSERT(class_type->name());
+  auto name = class_type->name()->qualifiedName();
+  TORCH_CHECK(!customClasses().count(name))
+  customClasses()[name] = std::move(class_type);
+}
 
-at::TypePtr noOpGetter(const std::string& /*unused*/) {
-  return nullptr;
+at::ClassTypePtr getCustomClass(const std::string& name) {
+  return customClasses().count(name) ? customClasses()[name] : nullptr;
 }
 
-std::atomic<GetCustomClassFnType> custom_class_fn{noOpGetter};
+bool isCustomClass(const c10::IValue& v) {
+  return v.isObject() && v.toObject()->type()->name() &&
+      getCustomClass(v.toObject()->type()->name()->qualifiedName());
+}
 
-}  // namespace
+std::unordered_map<std::string, std::shared_ptr<Function>>& customClassMethods() {
+  static std::unordered_map<std::string, std::shared_ptr<Function>>
+      customClassMethods;
+  return customClassMethods;
+}
 
-void setGetCustomClassFn(GetCustomClassFnType fn) {
-  custom_class_fn.store(fn);
+Function* getCustomClassMethod(const std::string& qualname) {
+  TORCH_CHECK(customClassMethods().count(qualname));
+  return customClassMethods()[qualname].get();
 }
 
-at::TypePtr getCustomClass(const std::string& name) {
-  return custom_class_fn.load()(name);
+void registerCustomClassMethod(std::shared_ptr<Function> fn) {
```

Comment:
Can we just stash them in a flat list to keep them alive? I want to avoid giving the impression that someone should look at this table to look up a method. They should go directly to the table in ClassType.